<table align="left" width=100%>
    <tr>
        <td width="20%">
            <img src="faculty.png">
        </td>
        <td>
            <div align="center">
                <font color="#21618C" size=8px>
                  <b> Praveen Faculty Notebook <br> (Day 1) </b>
                </font>
            </div>
        </td>
    </tr>
</table>

## Table of Content

1. **[What is Exploratory Data Analysis?](#eda)**
2. **[Working with a Dataset](#data)**
    - 2.1 - **[Read the Data](#read_data)**
    - 2.2 - **[Variable type](#variable)**
3. **[Measures of Central Tendency](#mct)**
    - 3.1 - **[Mean](#mean)**
    - 3.2 - **[Trimmed Mean](#tmean)**
    - 3.3 - **[Median](#median)**
    - 3.4 - **[Mode](#mode)**
    - 3.5 - **[Value Counts](#vcounts)**
    - 3.6 - **[Partition Values](#pvalues)**
4. **[Measures of Dispersion](#dispersion)**    
    - 4.1 - **[Range](#range)**
    - 4.2 - **[Variance](#variance)**
    - 4.3 - **[Standard Deviation](#std)**
    - 4.4 - **[Coefficient of Variation](#cv)**
    - 4.5 - **[Interquartile Range](#iqr)**
5. **[Distribution of the data](#distribution)** 
6. **[Shape of the data](#shape)** 
    - 6.1 - **[Skewness](#skewness)**
    - 6.2 - **[kurtosis](#kurtosis)**
7. **[Correlation and Covariance](#corr_cov)**  
    - 7.1 - **[Correlation](#corr)**
    - 7.2 - **[Covariance](#cov)**
    


<a id="eda"></a>
## 1. What is Exploratory Data Analysis?

Exploratory data analysis is an important step in any research analysis. It searches for patterns and relationships in the data. In other words, we may explain EDA as a process to summarise the crucial characteristics of the data to get better insights.<br>
We explore each dataset using multiple techniques and compare the results. We find answers to questions such as:<br>
&nbsp;&nbsp;&nbsp;      1) What is the shape of the data? <br>
&nbsp;&nbsp;&nbsp;      2) Are there redundant variables? <br>
&nbsp;&nbsp;&nbsp;      3) Are there missing or null values? <br>
&nbsp;&nbsp;&nbsp;      4) What is the average value of a feature? <br>
&nbsp;&nbsp;&nbsp;      5) Are there any outliers present? <br>
&nbsp;&nbsp;&nbsp;      6) Is there any relationship between these variables?

**Note:** The questions mentioned above are **not** the only questions answered in EDA. These are a few questions that one may start initially.

**Data analysis is divided into the following parts:**
<img src="analysis_types.png">

### 1. Descriptive Statistics:

Descriptive statistics summarize a given data set. It includes measures of central tendency, measures of variation and distribution of the data.
<br> The **measures of central tendency** include the mean, median, and mode, while **measures of variability** include the standard deviation, variance, the coefficient of variation, and so on.

### 2. Inferential Statistics:
Inferential statistics are procedures that help in making inferences about a population based on findings from a sample.<br>
It includes **testing of hypothesis** and **parameter estimation**. It allows us to infer parameters of the population based on sample and build models on it.


### What are Population and Sample?
The population is a complete set of elements having some common characteristics selected for a research study.<br>
A sample is a set of randomly chosen observations from a population. Samples are drawn from the population when a dataset is large since it is challenging to study the whole population. We draw a sample when the item under study is feared to exhaust. For instance,<br> 
<ul>
    <li>the study of the durability of an electric bulb </li>
        <li>the study of the effectiveness of new medicine</li>
</ul> 

**In this session, we will be using the `k-circlesales` dataset. We shall conduct exploratory data analysis by answering a few of the questions mentioned above. However, our primary focus will be on descriptive statistics.**

<a id="data"></a>
## 2. Working with a Dataset

The data definition which comes with the data is as follows:

**Item_Identifier:** Unique product ID assigned to every distinct item

**Item_Weight:** Weight of the product

**Item_Fat_Content:** Describes whether the product is low in fat or not

**Item_Visibility:** Total display area allocated to the particular product (in %)

**Item_Type:** Describes the food category to which the item belongs

**Item_MRP:** Maximum Retail Price (list price) of the product

**Outlet_Identifier:** Unique store ID assigned. It consists of an alphanumeric string of length 6

**Outlet_Establishment_Year:** The establishment year of the store

**Outlet_Size:** Size of the store in terms of ground area covered

**Outlet_Location_Type:** Size of the city in which the store is located

**Outlet_Type:** Is the outlet just a grocery store or a supermarket

**Profit:** Profit of the item sold (in %)

**Item_Outlet_Sales**: Sales of the product

<a id="read_data"> </a>
## 2.1 Read the data

We are using the `k-circlesales` dataset. Let us explore the data with some statistical metrics.


In [1]:
# import pandas
import pandas as pd

# read the data
df_sales = pd.read_csv('k_circle_sales.csv')
# check first five rows of data
df_sales.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy     249.8            OUT049   
1            Soft Drinks      48.3            OUT018   
2                   Meat     141.6            OUT049   
3  Fruits and Vegetables     182.1            OUT010   
4              Household      53.9            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 2   
1                       2009      Medium               Tier 2   
2                       1999      Medium               Tier 2   
3                       1998         NaN                  NaN   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  Profit  
0  Supermarket Type1          3735.1380    11.5  
1  Supermarket Type2           443.4228    14.3  
2  Supermarket Type1          2097.2700    14.5  
3      Grocery Store           732.3800    13.6  
4  Supermarket Type1           994.7052    14.1

In [2]:
df_sales.describe()

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  7774.000000      8523.000000  8523.000000                8523.000000   
mean     11.676740         0.066132   140.998838                1997.831867   
std       5.776851         0.051598    62.258099                   8.371760   
min       0.000000         0.000000    31.300000                1985.000000   
25%       7.720000         0.026989    93.800000                1987.000000   
50%      11.800000         0.053931   142.700000                1999.000000   
75%      16.500000         0.094585   185.650000                2004.000000   
max      21.350000         0.328391   266.900000                2009.000000   

       Item_Outlet_Sales       Profit  
count        8523.000000  8523.000000  
mean         2181.288914    13.414514  
std          1706.499616     1.701840  
min            33.290000     0.100000  
25%           834.247400    13.150000  
50%          1794.331000    13.900000  
75%          3101.296400    14.300000  
max         13086.964800    24.000000

In [3]:
df_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 13 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7774 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       6473 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
 12  Profit                     8523 non-null   float64
dtypes: float64(5), int64(1), object(7)
memory usage:

In [5]:
df_sales.columns

Index(['Item_Identifier', 'Item_Weight', 'Item_Fat_Content', 'Item_Visibility',
       'Item_Type', 'Item_MRP', 'Outlet_Identifier',
       'Outlet_Establishment_Year', 'Outlet_Size', 'Outlet_Location_Type',
       'Outlet_Type', 'Item_Outlet_Sales', 'Profit'],
      dtype='object')

Let us now see the number of variables and observations  in the data.

In [4]:
# use 'shape' attribute of the DataFrame to check the dimension of data
df_sales.shape

(8523, 13)

In [ ]:
df_sales.info()

In [6]:
df_sales['Item_Fat_Content'].unique()

array(['Low Fat', 'Regular', 'low fat', 'LF', 'reg'], dtype=object)

<b>Interpretation:</b> The data has 8523 rows and 13 columns that is there are 8523 observations and 13 variables   

<a id="variable"> </a>
## 2.2 Variable type

Identifying the type of variable we are working with is always the first step of the data analysis process. We classify the variables as follows:
<ol>
    <li><b>Numerical (quantitative)</b></li>
    <ul><li>Continuous - A random variable which takes continuous values, i.e. values within a range, usually measured, such as height (within the range of human heights)</li>
        <li>Discrete -  A random variable which takes discrete values, i.e. values from the set of whole numbers only. It can take countably finite values</li></ul><br>
    <li><b>Categorical (qualitative)</b></li>
    <ul>
        <li> Ordinal - Values that have inherent ordering levels, such as (high, medium, low)</li>
        <li>Nominal - Nominal data has no order. The values represent discrete units and are used to label variables (male or female)</li></ul>
</ol>  
Data has categorical and/or numerical variables. The main data types stored in pandas dataframes are object, float, int, boolean and datetime64. It is important to check whether the data types are correctly identified by python.<br>

In [7]:
# check the data type
df_sales.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
Profit                       float64
dtype: object

<b>Interpretation:</b> Refer to the data definition and check whether the data has been correctly identified.
Now based on the result, classify the various data columns into numerical/categorical attributes:<br>   <ul> 
                            <li> Numerical Attributes: Item_Weight,  Item_Visibility,  Item_MRP,  Outlet_Establishment_Year,  Item_Outlet_Sales and  Profit </li>
                            <li>Categorical Attributes: Item_Identifier, Item_Fat_Content, Item_Type, Outlet_Identifier, Outlet_Size, Outlet_Location_Type and Outlet_Type  </li>
    </ul>

<a id="central_tendency"> </a>
## 3. Measures of Central Tendency

Let us now understand the measures of central tendency.

<a id="mean"></a>
### 3.1 Mean
<ul>
  <li>Mean is the average of the data</li>
  <li>The important property of the mean is that it includes all values in the data. Thus it can be affected by the outliers. Presence of outliers can give unreliable results    
</ul>
<b>The questions we can answer now are:<br>
&nbsp;&nbsp; 1) What is the average profit earned for an item? <br>
&nbsp;&nbsp; 2) What is the average weight for the dairy items sold?
</b><br>

The answer to the questions above is to find the mean. Let us use the function `.mean()`

In [8]:
# calculate the average values
print("The mean values for the numeric variables in the data are:\n", df_sales.mean())

The mean values for the numeric variables in the data are:
 Item_Weight                    11.676740
Item_Visibility                 0.066132
Item_MRP                      140.998838
Outlet_Establishment_Year    1997.831867
Item_Outlet_Sales            2181.288914
Profit                         13.414514
dtype: float64


C:\Users\locha\AppData\Local\Temp/ipykernel_8564/1419692189.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  print("The mean values for the numeric variables in the data are:\n", df_sales.mean())


**Interpretation:**  The result shows the average for all the variables. We see the average weight of the item sold is 11.67 units, however not much can be inferred by this. The average sales and profit of an item in the data are 2181.288 INR and 13.41%  respectively. <br><br>
                        To answer the given questions:
                        <ol>
                            <li>The average profit earned on each item is 13.41%</li>
                        </ol>
                    </b>
 <b>However, to answer the second question, we need to first select the dairy items and then take the average of their weight  </b>

In [9]:
df_sales.columns

Index(['Item_Identifier', 'Item_Weight', 'Item_Fat_Content', 'Item_Visibility',
       'Item_Type', 'Item_MRP', 'Outlet_Identifier',
       'Outlet_Establishment_Year', 'Outlet_Size', 'Outlet_Location_Type',
       'Outlet_Type', 'Item_Outlet_Sales', 'Profit'],
      dtype='object')

In [10]:
df_sales.head()


Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy     249.8            OUT049   
1            Soft Drinks      48.3            OUT018   
2                   Meat     141.6            OUT049   
3  Fruits and Vegetables     182.1            OUT010   
4              Household      53.9            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 2   
1                       2009      Medium               Tier 2   
2                       1999      Medium               Tier 2   
3                       1998         NaN                  NaN   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  Profit  
0  Supermarket Type1          3735.1380    11.5  
1  Supermarket Type2           443.4228    14.3  
2  Supermarket Type1          2097.2700    14.5  
3      Grocery Store           732.3800    13.6  
4  Supermarket Type1           994.7052    14.1

In [11]:
df_sales[ df_sales['Item_Type'] == 'Dairy' ]['Item_Weight'].mean()

12.256701612903253

In [12]:
# select the rows having dairy item 
# declare it as 'df_dairy'
df_dairy = df_sales[df_sales.Item_Type == "Dairy"]

# obtain the average of its weight

df_dairy.Item_Weight.mean()

12.256701612903253

<b>Interpretation:</b> 2. The average weight of the dairy items sold is 12.25 units
                

<a id="tmean"> </a>
### 3.2 Trimmed Mean
We know that the mean value is affected by the extreme observations, so let us obtain the trimmed mean.

In [13]:
# The function to obtain the trimmed mean is available in the stats module of the library scipy
# The function computes the trimmed mean of a set of observations, we cannot pass the DataFrame
# First import the required libraries
# Then we  write a function 'trimmed_mean' to the get the trimmed mean all for numeric variable
# Also add a column for mean to aid comparison

# The user defined function:
# Step 1: Select the columns with numeric variable type
# Step 2: Create 3 empty lists for variable name, mean and trimmed mean
# Step 3: Using a for loop to compute the mean values
# Step 4: Append these values to the lists
# Step 5: Create a DataFrame from these lists

########################################################################################################

# import the libraries numpy and scipy 
# import the stats module from the scipy 
import numpy as np
import scipy
from scipy import stats


# function to obtain the trimmed mean
def trimmed_mean(data):
    num = pd.DataFrame(data.select_dtypes(include = np.number))    # select the columns with numeric variable type 
    n,mv, tmv = [],[],[]                                           # create 3 empty lists for variable name, mean and trimmed mean
    for i in range(num.shape[1]) :                                 # use a for loop 
        n.append(num.iloc[:,i].name)                               # using a for loop append the variable name                                             
        mv.append(num.iloc[:,i].mean())                            # using a for loop compute the mean values
        # proportiontocut: fraction of data points to cut off of both tails of the distribution
        tmv.append(scipy.stats.trim_mean(num.iloc[:,i], proportiontocut = 0.20)) # using a for loop compute the trimmed mean values
   
    df = pd.DataFrame([n,mv,tmv],index=["Variable","Mean","Trimmed Mean"])       # create a DataFrame from these lists
   
    return(df.T)                                                   # return the transpose of the DataFrame

In [14]:
# use the above defined function for our data
trimmed_mean(df_sales)

Variable         Mean Trimmed Mean
0                Item_Weight     11.67674    12.815851
1            Item_Visibility     0.066132     0.057186
2                   Item_MRP   140.998838   139.628074
3  Outlet_Establishment_Year  1997.831867  1998.595308
4          Item_Outlet_Sales  2181.288914  1883.197923
5                     Profit    13.414514    13.841134

**Interpretation:** The trimmed mean for outlet sales is lower than the mean value. We infer that there are significantly many extreme observations. Similarly, for the weight of products, there is a slight difference in the means, indicating the presence of outliers.
<br> There is not much difference in the values for profit and visibility of the item
                 

<a id="median"></a>
### 3.3 Median
<ul>
  <li>The median value is the middlemost value of the dataset, it divides the data into two equal halves. Thus, it is a positional value</li>
     <li>It can be obtained for categorical variables</li>
  <li>Unlike mean, the median value is not affected by extreme values</li>
    <li>The median value is at times used to impute the missing values in the data</li>
</ul>

In [15]:
# calculate the median
print("The median values for the numeric variables are:\n", df_sales.median())

The median values for the numeric variables are:
 Item_Weight                    11.800000
Item_Visibility                 0.053931
Item_MRP                      142.700000
Outlet_Establishment_Year    1999.000000
Item_Outlet_Sales            1794.331000
Profit                         13.900000
dtype: float64


C:\Users\locha\AppData\Local\Temp/ipykernel_8564/520938650.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  print("The median values for the numeric variables are:\n", df_sales.median())


**Interpretation:** The above values show the median for all the variables. These values divide the observations into two halves implying that there are exactly 50% of observations above this point and 50% below.
                        <br> For instance, consider the MRP of items, there are 50% of items with price less than 142.7 and 50% of items with price more than 142.7.

<a id="mode"></a>
### 3.4 Mode
<ul>
  <li> Mode of the data is the value which has the highest frequency</li>
  <li>It is a measure that can be used for categorical variables. Example: gender</li>
</ul>  

In [17]:
df_sales.select_dtypes(include=object)

Item_Identifier Item_Fat_Content              Item_Type  \
0              FDA15          Low Fat                  Dairy   
1              DRC01          Regular            Soft Drinks   
2              FDN15          Low Fat                   Meat   
3              FDX07          Regular  Fruits and Vegetables   
4              NCD19          Low Fat              Household   
...              ...              ...                    ...   
8518           FDF22          Low Fat            Snack Foods   
8519           FDS36          Regular           Baking Goods   
8520           NCJ29          Low Fat     Health and Hygiene   
8521           FDN46          Regular            Snack Foods   
8522           DRG01          Low Fat            Soft Drinks   

     Outlet_Identifier Outlet_Size Outlet_Location_Type        Outlet_Type  
0               OUT049      Medium               Tier 2  Supermarket Type1  
1               OUT018      Medium               Tier 2  Supermarket Type2  
2               OUT049      Medium               Tier 2  Supermarket Type1  
3               OUT010         NaN                  NaN      Grocery Store  
4               OUT013        High               Tier 3  Supermarket Type1  
...                ...         ...                  ...                ...  
8518            OUT013        High               Tier 3  Supermarket Type1  
8519            OUT045         NaN                  NaN  Supermarket Type1  
8520            OUT035       Small                Tier1  Supermarket Type1  
8521            OUT018      Medium               Tier 2  Supermarket Type2  
8522            OUT046       Small                Tier1  Supermarket Type1  

[8523 rows x 7 columns]

In [20]:
df_sales.select_dtypes(include=object)['Item_Identifier'].unique()

array(['FDA15', 'DRC01', 'FDN15', ..., 'NCF55', 'NCW30', 'NCW05'],
      dtype=object)

In [21]:
df_sales.select_dtypes(include=object)['Item_Identifier'].value_counts()

FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: Item_Identifier, Length: 1559, dtype: int64

In [22]:
df_sales.select_dtypes(include=object).value_counts()

Item_Identifier  Item_Fat_Content  Item_Type              Outlet_Identifier  Outlet_Size  Outlet_Location_Type  Outlet_Type      
DRA12            LF                Soft Drinks            OUT035             Small        Tier1                 Supermarket Type1    1
FDV27            Regular           Meat                   OUT019             Small        Tier1                 Grocery Store        1
FDV32            Low Fat           Fruits and Vegetables  OUT035             Small        Tier1                 Supermarket Type1    1
                                                          OUT019             Small        Tier1                 Grocery Store        1
                                                          OUT018             Medium       Tier 2                Supermarket Type2    1
                                                                                                                                    ..
FDJ33            Regular           Snack Foods            OU

In [23]:
df_sales.select_dtypes(include=object)['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

In [16]:
# filter the categorical variables from the data
df_cat = df_sales.select_dtypes(include=object)

# calculate the mode
df_cat.mode()

Item_Identifier Item_Fat_Content              Item_Type Outlet_Identifier  \
0           FDG33          Low Fat  Fruits and Vegetables            OUT027   
1           FDW13              NaN                    NaN               NaN   

  Outlet_Size Outlet_Location_Type        Outlet_Type  
0      Medium               Tier 2  Supermarket Type1  
1         NaN                  NaN                NaN

**Interpretation:** We see that the variable containing a unique product ID is bimodal (having two modes) and others are unimodal (having one mode). 

**From the above output, we see that the frequency of each observation is not given. In order to get a better insight, use the `describe()`.**

In [24]:
# get the summary of categorical variables
df_cat.describe()

Item_Identifier Item_Fat_Content              Item_Type  \
count             8523             8523                   8523   
unique            1559                5                     16   
top              FDW13          Low Fat  Fruits and Vegetables   
freq                10             5089                   1232   

       Outlet_Identifier Outlet_Size Outlet_Location_Type        Outlet_Type  
count               8523        6113                 6473               8523  
unique                10           3                    8                  4  
top               OUT027      Medium               Tier 2  Supermarket Type1  
freq                 935        2793                 2793               5577

<b>The output gives:
 <ol>
                            <li> count: number of values other than NaN </li>
                            <li> unique: number of unique values present in the variable</li>
                            <li> top: most repeated value   
                            <li> frequency: frequency of most repeated value</li>
                        </ol> 
</b><br>
**Interpretation:** From the output, we see that the variable representing the outlet size has missing values since its count 6113 is less than 8523 (number of observations in the data). Similarly, for variable representing type of outlet location there 6473 observations indicating missing values.
<br>There are 1559 unique items under study. Also, note here item FDG33 is the most common item with a frequency of 10. However, the previous output for mode it is observed that item FWD13 is also a common item with the same frequency- 10. It is a good practice to check for mode values using the mode() 
                     </b>
               

 <b>Note that the DataFrame we considered has categorical variables only. If we consider a DataFrame which has both categorical and numerical variables, then the describe function will give a summary of the numeric variables only. In order to get summary for categorical variables use describe(include="object")</b>

In [25]:
# summary of categorical values of a DataFrame
df_sales.describe(include="object") 

Item_Identifier Item_Fat_Content              Item_Type  \
count             8523             8523                   8523   
unique            1559                5                     16   
top              FDW13          Low Fat  Fruits and Vegetables   
freq                10             5089                   1232   

       Outlet_Identifier Outlet_Size Outlet_Location_Type        Outlet_Type  
count               8523        6113                 6473               8523  
unique                10           3                    8                  4  
top               OUT027      Medium               Tier 2  Supermarket Type1  
freq                 935        2793                 2793               5577

In [26]:
df_sales.describe() 

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  7774.000000      8523.000000  8523.000000                8523.000000   
mean     11.676740         0.066132   140.998838                1997.831867   
std       5.776851         0.051598    62.258099                   8.371760   
min       0.000000         0.000000    31.300000                1985.000000   
25%       7.720000         0.026989    93.800000                1987.000000   
50%      11.800000         0.053931   142.700000                1999.000000   
75%      16.500000         0.094585   185.650000                2004.000000   
max      21.350000         0.328391   266.900000                2009.000000   

       Item_Outlet_Sales       Profit  
count        8523.000000  8523.000000  
mean         2181.288914    13.414514  
std          1706.499616     1.701840  
min            33.290000     0.100000  
25%           834.247400    13.150000  
50%          1794.331000    13.900000  
75%          3101.296400    14.300000  
max         13086.964800    24.000000

**The output is the same as the above result.**

<a id="vcounts"></a>
### 3.5 Value Counts
<ul>
  <li>It is not possible to find the mean of a categorical variable</li>
  <li>However, it is possible to know the number of times a category appears in the data</li>
</ul> 

Let us note the categorical variables in the data.

In [27]:
# display the columns names of the DataFrame use the columns attribute
df_cat.columns

Index(['Item_Identifier', 'Item_Fat_Content', 'Item_Type', 'Outlet_Identifier',
       'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object')

The categorical variables in the data are Item_Identifier, Item_Fat_Content, Item_Type, Outlet_Identifier, Outlet_Size, Outlet_Location_Type, and Outlet_Type.

**Let us get a count of all the categories in the data.**

1. The variable `Item_Identifier`

In [28]:
# use value_counts() to get frequencies of each category
df_cat.Item_Identifier.value_counts()

FDW13    10
FDG33    10
NCY18     9
FDD38     9
DRE49     9
         ..
FDY43     1
FDQ60     1
FDO33     1
DRF48     1
FDC23     1
Name: Item_Identifier, Length: 1559, dtype: int64

**Interpretation:** As per the data definition, the variable 'Item_Identifier' is a unique product ID assigned to every distinct item. From the output, we can see there are 1559 items. The variable is bimodal since categories 'FDG33' and 'FDW13' have equal frequencies. This was inferred before when we used the mode() and describe()
                   

2. The variable `Item_Fat_Content`

In [29]:
# use the value_counts() to get frequencies of each category
df_cat.Item_Fat_Content.value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

**Interpretation:** We see that there are 5 categories viz, Low Fat, Regular, LF, reg, low fat. 
However, note that 'Low Fat' has also been referred to as 'low fat' and 'LF' in the dataset, and 'Regular'  as 'reg'
                    <br><br>
**We shall now rectify the inconsistencies using the replace()**

In [30]:
# to replace "low fat" by "Low Fat"
df_sales.replace(to_replace="low fat", value="Low Fat", inplace=True)

# to replace "low fat" by "Low Fat"
df_sales.replace(to_replace="LF", value="Low Fat", inplace=True)

# to replace "reg" by "Regular"
df_sales.replace(to_replace="reg", value="Regular", inplace=True)

In [31]:
# check whether the values have got replaced
# use value_counts() to get frequencies of each category
df_sales.Item_Fat_Content.value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

**Interpretation:** There are two categories present in the variable. The data has more items with less fat content
                   

3. The variable `Item_Type`

In [32]:
# use value_counts() to get frequencies of each category
df_cat.Item_Type.value_counts()

Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64

**Interpretation:** There are 16 varieties of items present in the data. Fruits and Vegetables seem to have the highest frequency. And seafood has the lowest frequency.

4. The variable `Outlet_Identifier`

In [33]:
# use value_counts() to get frequencies of each category
df_cat.Outlet_Identifier.value_counts()

OUT027    935
OUT013    932
OUT049    930
OUT046    930
OUT035    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
OUT019    528
Name: Outlet_Identifier, dtype: int64

**Interpretation:** There are 10 stores considered for the analysis. All stores have almost same frequency except for OUT010 and OUT019 have low observations.

5. The variable `Outlet_Size`

In [34]:
# use value_counts() to get frequencies of each category
df_cat.Outlet_Size.value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

**Interpretation:** This variable gives the size of the outlet. There are three categories-Small, Medium and High. There are relatively a lesser number of outlets which are classified as 'High'. The data comprises more of 'Small' and 'Medium' sized outlets. However, note that the data has missing values.

6. The variable `Outlet_Location_Type`

In [35]:
# use value_counts() to get frequencies of each category
df_cat.Outlet_Location_Type.value_counts()

Tier 2    2793
Tier1     2388
Tier 3     932
?          120
  --       109
  -         67
na          48
NAN         16
Name: Outlet_Location_Type, dtype: int64

**Interpretation:** The variable represents the size of the city in which the store is located. The city size has three categories. The 'Tier 2' city has the highest number of stores. However, there are other values such as `?`, `-`, `na`, which all represent missing data. We shall deal with missing data in later sessions.

<a id="pvalues"></a>
### 3.6 Partition Values (Quantiles)
<ul>
  <li>Partition values are the  values which divide the dataset into equal parts</li>
  <li>We have seen that the median divides the data into two equal halves. Similarly, in order to divide the data into four parts we use quartiles, for ten parts we use deciles</li>
</ul> 

**1. Quartiles:** 

Quartiles are the data points which partitions the data into 4 equal parts. So, there are 3 quartiles.

**The First Quartile:** The first quartile is the point that partitions the first 25% of the data and later 75%. This means that 25% of the data are less than or equal to the first quartile and 75% of the observations are greater than or equal to the first quartile. The first quartile is the same as the 25th percentile.


To get the quartiles we use the quantile function.

In [ ]:
# the first quartile
df_sales.quantile(0.25)

**The Second Quartile:** The second quartile is the point that partitions the data in two equal parts. This means that 50% of the data are less than or equal to the second quartile and 50% of the observations are greater than or equal to the second quartile. The second quartile is the same as the 50th percentile or median.


In [ ]:
# the second quartile
df_sales.quantile(0.5)

**The Third Quartile:** The third quartile is the point that partitions the data in 75-25 ratio. This means that 75% of the data are less than or equal to the third quartile and 25% of the observations are greater than or equal to the third quartile. The third quartile is called the 75th percentile.



In [ ]:
# the third quartile
df_sales.quantile(0.75)

**Now let us plot the three quartiles for the variable `Item_Outlet_Sales` in the data.**

In [ ]:
# import the library matplotlib.pyplot 
import matplotlib.pyplot as plt

# using the below code line, matplotlib graphs are included in the python notebook
%matplotlib inline

# set the plot size
plt.rcParams["figure.figsize"] = [15,5]

# compute the quartiles
q1 = df_sales.Item_Outlet_Sales.quantile(q=0.25)   # compute the first quartile
q2 = df_sales.Item_Outlet_Sales.quantile(q=0.50)   # compute the second quartile
q3 = df_sales.Item_Outlet_Sales.quantile(q=0.75)   # compute the third quartile

# plot the density plot the density plot
df_sales.Item_Outlet_Sales.plot(kind='density') 

# to plot the quartile lines
# plt.axvline(): plots a line perpendicular to the X-axis at specified location on X-axis
# x: location on X-axis at which the line is to be plotted
# ymax: specify the height of the line
# color: color of the line
# linestyle: specify the style of the line 
# label: the name of the line. This name is displayed in the legend
plt.axvline(x = q1, ymax=0.92, color='red', linestyle='--', label ="First Quartile")     # plot the first quartile
plt.axvline(x = q2, ymax=0.79, color='blue', linestyle='--', label ="Second Quartile")   # plot the second quartile
plt.axvline(x = q3, ymax=0.53, color='green', linestyle='--', label ="Third Quartile")   # plot the third quartile

# to display the legend using plt.legend()
# loc: specify the location of the legend
# fontsize: set the font size of the legends
plt.legend(loc="best", fontsize= 15)

# to display the plot title using plt.title()
# fontsize: set the font size of the title
plt.title("The outlet sales", fontsize = 20)

# to label the X-axis using plt.xlabel()
# fontsize: set the font size of the label
plt.xlabel("Item_Outlet_Sales", fontsize = 15)

# to label the Y-axis
# fontsize: set the font size of the label
plt.ylabel("Density", fontsize = 15)


# label the X-axis at the points where we draw the perpendicular lines using plt.xticks()
# locs: specify the location on X-axis
# labels: specify the labels on X-axis
# fontsize: set the font size of the ticks
locs,labels = [q1,q2,q3],["Q1","Q2","Q3"]           # get locations and labels
plt.xticks(locs, labels, fontsize = 15 )            # set locations and labels

# to display the figure
plt.show()

Similarly, you can draw a plot for other numeric variables in the data.

**2. Percentiles:**

Percentiles are the data points which partitions the data into 100 equal parts. So, there are 99 percentiles.

The $25^{th}$ percentile is the same as the $1^{st}$ quartile.<br>
The $50^{th}$ percentile is the same as the $2^{nd}$ quartile which is the median.<br>
The $75^{th}$ percentile is the same as the $3^{rd}$ quartile.<br>

Let us find the $10^{th}$ percentiles for the numeric variables in the data.

In [ ]:
# obtain the 10th percentile
df_sales.quantile(0.10)

**Interpretation:** The $10^{th}$ percentile partitions the data such that 10% of the data is less than or equal to it and 90% of the data is greater than or equal to it.

Let us find the $45^{th}$ percentiles for all the numeric variables in the data.

In [ ]:
# obtain the 45th percentile
df_sales.quantile(0.45)

**Interpretation:** The $45^{th}$ percentile partitions the data such that 45% of the data is less than or equal to it and 55% of the greater than or equal to this observation.

**Remark:** The partition values are not necessarily equispaced.

<a id="dispersion"> </a>
## 4 Measure of Dispersion
<br>
<b><ul>
  <li>The measure of dispersion indicates the variability within a set of measures</li>
  <li>Less variability implies more data is closer to the central observations. More variability implies more data is spread further away from the center
</li>
  <li>The most popular variability measures are the range, interquartile range (IQR), variance, and standard deviation
</li>
</ul> </b> 

<a id="range"> </a>
### 4.1 Range
<ul>
  <li>The range represents the difference between the largest and the smallest points in the data</li>
    <li>In spite of being very easy to define and calculate, it is not a reliable measure</li>
    <li>It is highly affected by extreme values</li>
</ul>  


Let us define a function to find range.

In [ ]:
def range_obs(data):
    r = data.max()-data.min()
    return(r)

**Remark:** We can not name the function as `range`, since the `range()` already exists

We shall now consider a few examples

In [ ]:
quantity_1 = pd.Series([2,4,6,8,10])
range_obs(quantity_1)

In [ ]:
quantity_2 = pd.Series([2,3,6,9,10])
range_obs(quantity_2)

In [ ]:
quantity_3 = pd.Series([2,10])
range_obs(quantity_3)

In [ ]:
quantity_4 = pd.Series([102,106,110])
range_obs(quantity_4)

**Interpretation:** It is apparent from the above results that the range value is invariant of the number of observations and their magnitude.

However, we will obtain the range metric for the numeric variables in our dataset `k-circlesaless`.

In [ ]:
# filter the numeric variables from the data
df_num = df_sales.select_dtypes(include = np.number)

# find the range
print(df_num.max()-df_num.min())

**Interpretation:** The range for the variable `Item_Visibility ` is very small, we may conclude that the dispersion is also less. <br>
For the variable `Item_Outlet_Sales`, the range is 13053.67. We may say that there are outliers present which affect the range value.<br>
However, it is not wise to compare variation in variables having different units because they vary in their respective units and the means about which they occur. 


<a id="variance"> </a>
### 4.2 Variance
<ul>
  <li>Variance measure the dispersion of the data from the mean</li>
  <li>It is an average of the sum of squares of difference between an observation and the mean. Thus the variance is always positive</li>
    <li>The variance is based on all the observations</li>
</ul>

We shall now consider a few examples. We take the same observations we took to demonstrate range.

In [ ]:
quantity_1 = pd.Series([2,4,6,8,10])
quantity_1.var()

In [ ]:
quantity_2 = pd.Series([2,3,6,9,10])
quantity_2.var()

In [ ]:
quantity_3 = pd.Series([2,10])
quantity_3.var()

**Interpretation:** The range for all the three series is 8. However, the variance values are different. It is a better measure of variability than range. It tells how spread the data is. It is seen that the series `quantity_1` has the least variance. <br>

**Now we will see a variance of the dataset**

In [ ]:
# calculate the variance
var = df_sales.var()

# rounding off the values upto 3 decimals
round(var, 3)

**Interpretation:** The variance indicates how much is the data varying from the mean. The output suggests that the variable `Item_Visibility` has the least variance and the variable `Item_Outlet_Sales` has the maximum variance.<br>
However, it is not sensible to compare variation in variables having different units because they vary in their respective units and the means about which they occur.
<br><br>We know that the variance is the sum of squares of the difference, so <em>the unit of variance is the squared unit of the observations</em>. Hence it is viable to take the square root of the variance to get an answer back in the original unit of measurement.

<a id="std"> </a>
### 4.3 Standard Deviation

<ol>
    <li> Standard deviation is the positive square root of the variance </li>
    <li> It has the same unit as that of the observations</li>
    <li> Like variance, it also measures the variability of the observations from the mean </li>
    </ol>

In [ ]:
# calculate the standard deviation
print(df_sales.std())
print(df_sales.var())

**Interpretation:** We can see that the variable 'Item_Visibility' has a standard deviation near to zero. It indicates that there is not much variability. Thus this variable does not add much value to the analysis and can be further removed from the study. <br>
However, it is not sensible to compare variation in variables having different units because they vary in their respective units and the means about which they occur.

<a id="cv"> </a>
### 4.4 Coefficient of Variation
<ul>
    <li>The coefficient of variation is a statistical measure of dispersion of data points around the mean</li>
    <li>It is a unit free measure and is always expressed in percentage</li>
    <li>We compare the coefficient of variation in order to decide which of the two sets of observations has more spread</li>
    </ul>

**Consider the following scenario: <br><br>
The regional swimming tournament is going to commence next month. The school has to form a team. The performance scores for five swimmers from their last five competitions are known. Based on these scores, determine which two swimmers should be on the team? (The scores are measured out of 10)**

<table>
  <tr>
    <th>Swimmer</th>
    <th colspan="5" >Scores of previous competitions</th>
  </tr>
  <tr>
    <td>Anne</td>
    <td>4.5</td>
    <td>8.8</td>
    <td>8.9</td>
    <td>9.2</td>
    <td>8.1</td>     
  </tr>
  <tr>
    <td>John</td>
    <td>5.5</td>
    <td>9.7</td>
    <td>8.8</td>
    <td>6.4</td>
    <td>7.5</td>     
  </tr>
  <tr>
    <td>Sarah</td>
    <td>8.5</td>
    <td>7.5</td>
    <td>7.9</td>
    <td>8.8</td>
    <td>9.1</td>     
  </tr>
  <tr>
    <td>Steve</td>
    <td>9.2</td>
    <td>8.8</td>
    <td>9.5</td>
    <td>9.2</td>
    <td>8.9</td>     
  </tr>
  <tr>
    <td>Sam</td>
    <td>4.5</td>
    <td>9.0</td>
    <td>8.9</td>
    <td>5.2</td>
    <td>7.1</td>     
  </tr>  
</table>

In [ ]:
# create series of the scores for all the five swimmers
swimmer_anne = pd.Series([6.5, 8.8, 8.9, 9.2, 8.1]) 
swimmer_john = pd.Series([5.5, 9.7, 8.8, 6.4, 7.5]) 
swimmer_sarah = pd.Series([8.5, 7.5, 7.9, 8.8, 9.1]) 
swimmer_steve = pd.Series([9.2, 8.8, 9.5, 9.2, 8.9]) 
swimmer_sam = pd.Series([4.5, 5.0, 5.9, 5.2, 5.1]) 

In [ ]:
# obtain coefficient of variation for Anne's scores
print("The coefficient of variation for Anne is: ",scipy.stats.variation(swimmer_anne))   

# obtain coefficient of variation for John's scores
print("The coefficient of variation. for John is: ",scipy.stats.variation(swimmer_john))  

# obtain coefficient of variation for Sarah's scores
print("The coefficient of variation for Sarah is: ",scipy.stats.variation(swimmer_sarah))

# obtain coefficient of variation for Steve's scores
print("The coefficient of variation for Steve is: ",scipy.stats.variation(swimmer_steve)) 

# obtain coefficient of variation for Sam's scores
print("The coefficient of variation for Sam is: ",scipy.stats.variation(swimmer_sam))       

**Interpretation:** The coefficient of variation tells the measure of dispersion of data points around the mean. Lesser coefficient of variation, the better the data is. <br>On comparing the coefficient of variation of all the swimmers, it is apparent Sarah and Steve have the least variation. This means that their scores are consistently good. Thus, Sarah and Steve should be on the team.

Let us get the coefficient of variation for the numeric variables in our data.

In [ ]:
# importing variation from scipy.stats 
from scipy.stats import variation  
 
for i in df_num.columns:
    print(i,":", scipy.stats.variation(df_sales[i],nan_policy='omit'))

**Interpretation:** The coefficient of variation is used to compare two or more samples. This output is not meaningful since we can not compare any of the variable with each other to get insights.

<a id="iqr"> </a>
### 4.5 Interquartile Range (IQR)
<ul>
    <li>The interquartile range is another useful measure of spread based on the quartiles</li>
    <li>It is the difference between the third quartile and the first quartile</li>
    <li>The IQR gives the range of middle 50% of the data</li>
    <li> It also helps in identifying the outliers
    </ul>

In [ ]:
# filter the numeric variables from the data
df_num = df_sales.select_dtypes(include = np.number)

# find the range
print(df_num.quantile(0.75)-df_num.quantile(0.25))

**Interpretation:** The IQR gives the range in which the middle 50% of the data lies. The smaller the range is, lesser is the dispersion.

<a id="distribution"> </a>
## 5. Distribution of the data

<ul>
    <li>The distribution is a summary of the frequency of values taken by a variable</li>
    <li>The distribution of the data gives information on the shape and spread of the data</li>
    <li>On plotting the histogram or a frequency curve for a variable, we are  actually looking at how the data is distributed over its range</li> 


In [ ]:
# set plot size
plt.rcParams["figure.figsize"] = [15,10]

# plot density plot using plot()
# kind: set the type of plot
# subplots: indicates whether to plot subplot for each variable or a single line plot
# layout: specify the arrangement of the subplots
# sharex: indicates whether to have the same scale on x-axis of all subplots
df_sales.plot(kind = 'density', subplots = True, layout = (3,3), sharex = False)

# show plot
plt.show()

In [ ]:
# import the seaborn library
import seaborn as sns

# initiate the plot
# plot 6 subplots in 2 rows and columns
# figsize: set figure size
fig, ax = plt.subplots(2, 3, figsize=(15, 10))

# plot a count plot for categorical variables in the data
# use a for loop to plot the count plot for the categoric variable
for variable, subplot in zip(df_cat.columns[1:], ax.flatten()):     
    z = sns.countplot(df_sales[variable], ax=subplot)                 # plot the count plot
    z.set_xticklabels(z.get_xticklabels(),rotation="90")              # rotate the on x-axis labels
    z.set_xlabel(variable, fontsize = 15)                             # set the size of x-axis labels
    

# tight_layout() adjusts subplot parameters to give specified padding
# pad specifies the padding (space) between the edges of subplots
plt.tight_layout(pad=4.0)

# display the plot    
plt.show()

**Interpretation:** The above output gives the distribution of variables in the data. Notice the differences in scales for each of the variables. <br> 
We have not plotted a count plot for `Item_Identifier` since there are 1559 categories.<br><br>
To infer more on the distribution of the data we must study the shape of the data. The next section deals with the skewness and kurtosis of the data

<a id="shape"> </a>
## 6 Shape of the data

<a id="skewness"> </a>
### 6.1 Skewness
<ul>
  <li>Skewness helps us to study the shape of the data.</li>
  <li>It represents how much a distribution differs from a normal distribution, either to the left or to the right.</li>
    <li>The value of the skewness can be either positive, negative or zero.</li>
</ul>  

In [ ]:
# use skew() to calculate the skewness
skew = df_sales.skew()

# print the skewness
print(skew)

**Interpretation:** The above value represents that the variable `Item_Visibility` and `Item_Outlet_Sales` are highly positively skewed.<br>The variables `Item_Weight` and `Item_MRP` are near normal. <br>The variable `Profit` has negative skewness. <br>
**Note:** It is advisable to check the distribution plot before making any conclusion.                    

<a id="kurtosis"> </a>
### 6.2 Kurtosis
<ul>
    <li> Kurtosis measures the peakedness of the distribution</li>
   <li>In other words, kurtosis is a statistical measure that defines how the tails of the distribution differ from the normal distribution</li>
       <li>Kurtosis identifies whether the tails of a given distribution contain extreme values</li>
</ul>  

To find the kurtosis of the data, we find use the function `kurt()`

In [ ]:
# use kurt() to calculate the kurtosis
df_sales.kurt()

**Interpretation:** The above output represents that the variables `Item_MRP`  and `Outlet_Establishment_Year` are platykurtic.<br>
The variables `Item_Weight` are near normal.<br>
The variable `Profit` is leptokurtic.<br>
**Note:** It is advisable to check the distribution plot before making any conclusion. 

## Let us plot the mean, median and mode for skewed distributions

We define a function histogram as per our requirement.

In [ ]:
def histogram(var):
    plt.rcParams["figure.figsize"] = [15,5]         # set the figure size
    var.hist(color="steelblue", grid=False)         # plot histogram
    mu = var.mean()                                 # calculate the mean
    mo = var.mode()                                 # calculate the mode
    me = var.median()                               # calculate the median
    
    # plot the lines for mean, median, and mode
    # use the function axvline()
    # x: the point at which the line is to be plotted
    # color: color of the line 
    # linestyle: specify the style of the line 
    # label: the name of the line. This name is displayed in the legend
    plt.axvline(x = mu, color='red',linestyle='--',label="mean")
    plt.axvline(x = me, color='orange',linestyle='--',label="median")
    plt.axvline(x = mo[0], color='black',linestyle='--',label="mode")  # consider first mode (for multimodal distributions)
   
    # print the legend of the data    
    plt.legend(prop={'size': 20}) 

In [ ]:
# import the required libraries 
from matplotlib import gridspec

# set the plot size
plt.rcParams["figure.figsize"] = [25,10]

# specify the geometry of the grid that a subplot is placed in
# split the plot into 2 rows and 3 columns
gs = gridspec.GridSpec(2, 3, width_ratios=[2.5, 2.5, 2.5], height_ratios=[0.20,0.01])

# to draw the plots

# step 1: specify the plot location by calling 'gs' initiated above 
# step 2: plot the histogram by calling the above histogram function
# step 3: specify the plot title by set_title()
#         fontsize: set the font size of the title


# plot in 1st row and 1st column
a1 = plt.subplot(gs[0,0])             
a1.set_title("Negatively Skewed Distribution", fontsize = 25)   
histogram(df_sales['Profit'])       
plt.xlabel("Profit",fontsize=15)                                


# plot in 1st row and 2nd column
b1 = plt.subplot(gs[0,1])
b1.set_title("Not Skewed Distribution", fontsize = 25)
histogram(df_sales['Item_MRP'])       
plt.xlabel("Item_MRP",fontsize=15) 


# plot in 1st row and 3rd column

# specify the plot location by calling 'gs' initiated above 
c1 = plt.subplot(gs[0,2])
c1.set_title("Positively Skewed Distribution", fontsize = 25)
histogram(df_sales['Item_Outlet_Sales'])       
plt.xlabel("Item_Outlet_Sales",fontsize=15) 


# to display the text

# step 1: specify the plot location by calling 'gs' initiated above 
# step 2: write the text in the plot using text()
#         x: location on x-axis where the text the is to be written
#         y: location on y-axis where the text the is to be written
#         s: text to be written 
#         fontsize: set the font size
# step 3: use .axis('off') to hide the x and y axes
# 

# plot in 2nd row and 1st column
a2 = plt.subplot(gs[1, 0])                                      
a2.text(x = 0.1, y = 0.3, s="mean < median < mode", fontsize=25)  
a2.axis('off')

# plot in 2nd row and 2nd column
b2 = plt.subplot(gs[1, 1]) 
b2.text(x = 0.1, y = 0.3, s="mean = median = mode" , fontsize=25)
b2.axis('off')

# plot in 2nd row and 3rd column 
c2 = plt.subplot(gs[1, 2])
c2.text(0.1, 0.3, s="mean > median > mode", fontsize=25)
c2.axis('off')

# to display the plot
plt.show()

**Interpretation:** We verified the facts:
                        <ol>
                            <li> For a negatively skewed distribution, $mean < median < mode$ </li>
                            <li> For a distribution with no skewness, $mean = median = mode$ (approximately) </li>
                            <li> For a positively skewed distribution, $mean > median > mode$ </li>
                        </ol>
                   

#### We shall see how a boxplot looks for skewed distributions.

<a id="boxplot"> </a>
### Box-plot



<img src="Boxplot.png">

The box of the above boxplot gives middle 50% of the data. It is the IQR. The line in the box represents the median.

Now with the help of a distribution plot, boxplot, and the summary statistics of a variable let us understand the data.

In [ ]:
# import the required libraries 

import seaborn as sns

# set the plot size
plt.rcParams["figure.figsize"] = [20,15]

# divides the plot into the specified number of subplots 
gs = gridspec.GridSpec(3, 3, width_ratios=[2.5, 2.5, 2.5], height_ratios=[.20, .10, .15])
  
# DISTRIBUTION PLOT
# step 1: specify the location of the plot
# step 2: specify the title of the plot
# step 3: plot a distribution plot
# step 4: if required set the scale of X-axis

# for the variable Profit
a1 = plt.subplot(gs[0, 0])
a1.set_title("Negatively Skewed", fontsize=15)
a1= sns.distplot(df_sales.Profit)


# for the variable Item_MRP
b1 = plt.subplot(gs[0, 1])
b1.set_title("Not Skewed", fontsize=15)
b1= sns.distplot(df_sales.Item_MRP)

# for the variable Item_Outlet_Sales
c1 = plt.subplot(gs[0, 2])
c1.set_title("Positively Skewed", fontsize=15)
c1= sns.distplot(df_sales.Item_Outlet_Sales)

# BOXPLOT
# step 1: specify the location of the plot
# step 2: plot the box plot
# step 3: if required set the scale of X-axis using plt.xlim()

# for the variable Profit
a2 = plt.subplot(gs[1, 0])
a2= sns.boxplot(df_sales.Profit)
plt.xlim(0, 20)

# for the variable Item_MRP
b2 = plt.subplot(gs[1, 1])
b2= sns.boxplot(df_sales.Item_MRP)

# for the variable Item_Outlet_Sales
c2 = plt.subplot(gs[1, 2])
c2= sns.boxplot(df_sales.Item_Outlet_Sales)

# SUMMARY STATISTICS
# step 1: specify the location of the plot
# step 2: specify the text along with its location
# step 3: hide the x and y axes

# for the variable Profit
a3 = plt.subplot(gs[2, 0])
a3.text(0.05, 0.2, s=str(df_sales.Profit.describe()), fontsize=15)
a3.axis('off')

# for the variable Item_MRP
b3 = plt.subplot(gs[2, 1])
b3.text(0.05, 0.2, s=str(df_sales.Item_MRP.describe()), fontsize=15)
b3.axis('off')

# for the variable Item_Outlet_Sales
c3 = plt.subplot(gs[2, 2])
c3.text(0.05, 0.2, s=str(df_sales.Item_Outlet_Sales.describe()), fontsize=15)
c3.axis('off')

# display the plot
plt.show()

**Interpretation:**<ol> <li> For the negatively skewed distribution, we see that the median in the boxplot is towards the third quartile. It can be verified with the help of the summary statistics displayed below it</li>
                            <li> For a distribution with no skewness, we see that the median in the boxplot is equidistant from the first and the third quartiles </li>
                            <li> For the positively skewed distribution, we see that the median in the boxplot is towards the first quartile </li>
                            <li> The points beyond the whiskers of the boxplot are considered to be outliers</li>
                    </ol>

### Boxplot for outliers

A boxplot is also used to identify outliers. The points beyond, the whiskers are the outliers.

In [ ]:
# set the plot size
plt.figure(figsize=(15,5))

# plot the boxplot
sns.boxplot(df_sales.Profit)

# to draw the arrows 
# use the method annotate()
# annotate the text at point xy
# xytext: location of the text and the tail of the arrow
# xy: location of the text and the head of the arrow
# arrowprops: color of the arrow
plt.annotate("Outliers",xy = (7.5,0.01), xytext = (7.5, 0.3),arrowprops = {'facecolor' : 'black'})
plt.annotate("Outliers",xy = (17.5,0.1),  xytext = (17.5, 0.3),arrowprops = {'facecolor' : 'black'} )

plt.show()

**Interpretation:** From the above plot, it is seen that the variable `Profit` has outliers

**In the later session, we will learn more about outlier treatment.**

### Z-Score

Another way to detect outliers are using the z-score
<ul>
  <li>Z-score of a value is the difference between that value and the mean, divided by the standard deviation</li>
  <li>Z-score of 0 indicates the value is the same as the mean</li>
  <li>Z-score is positive or negative means the value is above or below the mean and by how many standard deviations</li>
  <li>Z-score greater than 3 or less than -3, indicates an outlier value</li>
</ul>

In [ ]:
# Z-scores are defined for each observation in a variable
# compute the z-scores using the method z-score from the scipy library 
zscores_Profit = scipy.stats.zscore(df_sales.Profit)

# print the z-score
zscores_Profit

The output gives the z-scores of observations in the variable `Profit`.

In [ ]:
# to get the outliers in the variable
zscores_Profit[(zscores_Profit < -3) | (zscores_Profit > 3)]

**Interpretation:** The above output gives the z-scores which are less than -3 and greater than 3. The values corresponding to these score are considered to be outliers

<a id="corr_cov"> </a>
## 7. Covariance and Correlation

<a id="corr"></a>
### 7.1 Correlation
  <li>It shows whether pairs of variables are related to each other</li>
  <li>If there is correlation, it shows how strong the correlation is</li>
  <li>Correlation takes values between -1 to +1, where values close to +1 represents strong positive correlation while values close to -1 represents strong negative correlation</li>
</ul>  


In [ ]:
# calculate the correlation between the variables
corr = df_sales.corr()

# print correaltion matrix
corr

In [ ]:
# set the plot size
fig,ax = plt.subplots(figsize=(10, 10))

# plot a heatmap for the correlation matrix
# annot: print values in each cell
# linewidths: specify width of the line specifying the plot
# vmin: minimum value of the variable
# vmax: maximum value of the variable
# cmap: colour code of the plot
# fmt: set the decimal place of annot
sns.heatmap(corr, annot = True, linewidths = 0.05, vmin = -1 , vmax = 1, cmap = "YlGnBu" , fmt = '10.4f')

# display the plot
plt.show()

<a id="cov"> </a>
## 7.2 Covariance

<ul>
  <li>It is the relationship between a pair of random variables where change in one variable causes change in another variable</li>
  <li>It can take any value between -infinity to +infinity, where the negative value represents the negative relationship whereas a positive value represents the positive relationship</li>
</ul>  

</body>
</html>

In [ ]:
# obtain the covariance matrix
cov = df_sales.cov()
cov

In [ ]:
# set the plot size
fig,ax = plt.subplots(figsize=(15, 8))

# plot a heatmap for the correlation matrix
# annot: print values in each cell
# linewidths: specify width of the line specifying the plot
# vmin: minimum value of the variable
# vmax: maximum value of the variable
# cmap: colour code of the plot
# fmt: set the decimal place of annot
sns.heatmap(cov, annot = True, linewidths = 0.05, cmap = "YlGnBu", fmt = '.4f')

# display the plot
plt.show()

In this session we understood the variables in our dataset. We have a fair idea regarding its distribution and spread. In the next sessions, we shall continue with the univariate analysis followed by multivariate.

**Let us export this data to use for the next session.**

In [ ]:
df_sales.to_csv("k-circlesales-cleaned.csv")